maven-core/src/main/java/org/apache/maven/utils/dag/DAG.java

```diff
@@ -0,0 +1,239 @@
+
+            vertexList.add(retValue);
+        }
+
+        return retValue;
+    }
+
+    public void addEdge(final String from, final String to) throws CycleDetectedException {
+        final Vertex v1 = addVertex(from);
+
+        final Vertex v2 = addVertex(to);
+
+        addEdge(v1, v2);
+    }
+
+    public void addEdge(final Vertex from, final Vertex to) throws CycleDetectedException {
+
+        from.addEdgeTo(to);
+
+        to.addEdgeFrom(from);
+
+        final List<String> cycle = CycleDetector.introducesCycle(to);
+
+        if (cycle != null) {
+            // remove edge which introduced cycle
+
+            removeEdge(from, to);
+
+            final String msg = "Edge between '" + from + "' and '" + to + "' introduces to cycle in the graph";
+
+            throw new CycleDetectedException(msg, cycle);
+        }
+    }
+
+    public void removeEdge(final String from, final String to) {
+        final Vertex v1 = addVertex(from);
+
+        final Vertex v2 = addVertex(to);
+
+        removeEdge(v1, v2);
+    }
+
+    public void removeEdge(final Vertex from, final Vertex to) {
+        from.removeEdgeTo(to);
+
+        to.removeEdgeFrom(from);
+    }
+
+    public Vertex getVertex(final String label) {
+        final Vertex retValue = vertexMap.get(label);
+
+        return retValue;
+    }
+
+    public boolean hasEdge(final String label1, final String label2) {
+        final Vertex v1 = getVertex(label1);
+
+        final Vertex v2 = getVertex(label2);
+
+        final boolean retValue = v1.getChildren().contains(v2);
+
+        return retValue;
+    }
+
+    /**
+     * @param label see name
+     * @return the childs
+     */
+    public List<String> getChildLabels(final String label) {
+        final Vertex vertex = getVertex(label);
+
+        return vertex.getChildLabels();
+    }
+
+    /**
+     * @param label see name
+     * @return the parents
+     */
+    public List<String> getParentLabels(final String label) {
+        final Vertex vertex = getVertex(label);
+
+        return vertex.getParentLabels();
+    }
+
+    /**
+     * @see java.lang.Object#clone()
+     */
+    @Override
+    public Object clone() throws CloneNotSupportedException {
+        // this is what's failing..
+        final Object retValue = super.clone();
+
+        return retValue;
+    }
+
+    /**
+     * Indicates if there is at least one edge leading to or from vertex of given label
+     * @param label the label
+     * @return <code>true</code> if this vertex is connected with other vertex,<code>false</code> otherwise
+     */
+    public boolean isConnected(final String label) {
+        final Vertex vertex = getVertex(label);
+
+        final boolean retValue = vertex.isConnected();
+
+        return retValue;
+    }
+
+    /**
+     * Return the list of labels of successor in order decided by topological sort
+     *
+     * @param label The label of the vertex whose predecessors are searched
```

Comment:
nit: the

Reply:
@elharo 
Is the "The" missing in front of the `label`?